pkg/controllers/rollout/controller.go

```diff
@@ -244,43 +277,63 @@ func pickBindingsToRoll(allBindings []*fleetv1beta1.ClusterResourceBinding, late
 	canBeUnavailableBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
 
 	// Those are the bindings that are candidates to be updated to be bound during the rolling phase.
-	boundingCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+	boundingCandidates := make([]toBeUpdatedBinding, 0)
 
 	// Those are the bindings that are candidates to be removed during the rolling phase.
-	removeCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+	removeCandidates := make([]toBeUpdatedBinding, 0)
 
 	// Those are the bindings that are candidates to be updated to latest resources during the rolling phase.
-	updateCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+	updateCandidates := make([]toBeUpdatedBinding, 0)
 
 	// Those are the bindings that are a sub-set of the candidates to be updated to latest resources but also are failed to apply.
 	// We can safely update those bindings to latest resources even if we can't update the rest of the bindings when we don't meet the
 	// minimum AvailableNumber of copies as we won't reduce the total unavailable number of bindings.
-	applyFailedUpdateCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+	applyFailedUpdateCandidates := make([]toBeUpdatedBinding, 0)
 
 	// calculate the cutoff time for a binding to be applied before so that it can be considered ready
 	readyTimeCutOff := time.Now().Add(-time.Duration(*crp.Spec.Strategy.RollingUpdate.UnavailablePeriodSeconds) * time.Second)
 
 	// classify the bindings into different categories
 	// TODO: calculate the time we need to wait for the first applied but not ready binding to be ready.
+	crpKObj := klog.KObj(crp)
 	for idx := range allBindings {
 		binding := allBindings[idx]
+		desiredBinding := binding.DeepCopy()
```

Comment:
When there's not many bindings to actually update, this seems to be wasting a lot of cycles deep-copying objects and setting their fields, which are not going to be used at all.

Reply:
updated,  right now it will only build the updateInfo when the resources have been changed.